### ISSUE TITLE
The All-Seeing Eye & Gold Box Companion support

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#3106]
@Python-Exoproject reported that Staging doesn't work with [The All Seeing-Eye](https://ase.zorbus.net/) (ASE) and [Gold Box Companion](https://gbc.zorbus.net/) (GBC) like vanilla DOSBox and ECE do. These are Windows-only "companion apps" for the Eye of the Beholder and the SSI Gold Box RPG series, respectively, and eXoDOS has support for them in their user launch options. Most notably, they add automap support, quality-of-life enhancements and various hacks and cheats.

It needs further investigation of what is needed to support ASE and GBC in Staging. This potentially involves working together with the author of ASE and GBC (it's the same guy) and perhaps asking him to tweak his apps a bit to make them more "Staging friendly", but we'll see.

**Comment by MasterO2** [comment]
The email address of The All Seeing-Eye's creator is on that site, so contacting him shouldn't be difficult.

**Comment by kcgen** [comment]
The authors memory inspection tools (like cheat engines) found a hardcoded offset where they could read the DOS base address value from the original DOSBox and ECE (on Windows). 

They could do the same for Staging, but to make it even easier, we added logging of the DOS memory base address in release 0.76 (and backported to 0.75.1).

This approach is not only cross-platform, but robust again changes to the memory layout of DOSBox itself. 

So I'm not sure there's anything for us to do. The author can update their tool to parse our log and grab the base address.

**Comment by kcgen** [comment]
@Python-Exoproject, if we can make a counter-offer 😅, please ask ACE and GBC authors support macOS and Linux, too (or alternatively: open-source these tools so community members can make cross-platform ports). 

Windows-only tooling doesn't fly very far here 🛬

**Comment by johnnovak** [comment]
> They could do the same for Staging, but to make it even easier, we added logging of the DOS memory base address in release 0.76 (and backported to 0.75.1).
> 
> This approach is not only cross-platform, but robust again changes to the memory layout of DOSBox Staging itself.

I tried to paste in the logged based memory addresses into GBC manually, and the tool still did not work. Not saying Staging is necessarily at fault, maybe it's the tool, but I'd like to keep this ticket open until issue is resolved one way or another.

Just for practical reasons, I want to be able to say, "yes, Staging works with ASE and GBC", because right now I couldn't get them to work even after 30+ mins of tinkering, neither could @Python-Exoproject. They JustWork(TM) with ECE—we want the same experience. This prevents eXoDOS from upgrading to Staging in their normal launch method, if I understand correctly. So people can't enjoy CRT shaders and ASE / GBC at the same time.

I'm happy to tinker with this later and/or contact the author as I hold the EoB and SSI Gold Box series in very high regard.

**Comment by kcgen** [comment]
DOSBox and ECE aren't logging any DOS addresses, so whatever these things tie into, it's probably something those authors have hardcoded (as an offset).

We know the cheat engines work, so hopefully author(s) can connect the dots. (Just saying that the ball is 100% in their court, as we have no visibility into whatever they're using or doing.)

**Comment by kcgen** [comment]
> So people can't enjoy CRT shaders and ASE / GBC at the same time.

Alternative: roll up the sleeves and use GridMonger. Propper old school mapping! :medal_sports:

**Comment by Grounded0** [comment]
> Haha, ok, but 99% of our users are on Windows, and these tools just work with ECE. I want them to just work with Staging on Windows, whatever it takes.

I wish it was also 99% of the code contributions and money coming in.

**Comment by johnnovak** [comment]
I had a correspondence with the author on a public forum, and this is what he said about his tool:

> GBC doesn't just use offsets to identifiers to read / change stuff in the game's memory, but also uses pointers and linked lists. I tried some non-standard DOSBox at some point, and could identify the game's location in memory, but those pointers still were not at the expected places. Trying to fix those for all the 12 games is time consuming and so utterly boring that at the time I just gave up. And there are even different versions of the games floating around.

So basically the tools assume the presence of certain _emulator-side_ data structures; they not only rely on the emulated machine's memory. Well, that makes them virtually impossible to support on our side, the only fix is for the author to rethink his approach and rewrite the tools, but as per above, that's unlikely to happen (understandably).

**Comment by Grounded0** [comment]
Seems to be cross platform enough for me since Wine/CrossOver runs it fine:

<img width="825" alt="Screenshot 2023-11-25 at 9 01 51" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/b432816b-0530-4532-af8b-5995e70e00ea">

<img width="1115" alt="Screenshot 2023-11-25 at 9 06 39" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/07973033-1f85-4053-a2a2-cdbca5dc0bee">

I'm not sure does this require running Windows version of DOSBox to be compatible under macOS and Linux (probably) but that's an acceptable tradeoff.

Hopefully the author adds support for DOSBox Staging because DOSBox will not run even on Windows in a few years due to SDL1 and general bit rot.

**Comment by johnnovak** [comment]
> Hopefully the author adds support for DOSBox Staging because DOSBox will not run even on Windows in a few years due to SDL1 and general bit rot.

As per above. Maybe, but he kinda lost motivation.

**Comment by emxd** [comment]
While this probably won't help the author, I did a bisect. Starting from b26300533e052fc35ec8f72a224660742befe299 GBC is no longer compatible. Before that 32-bit works fine.

**Comment by johnnovak** [comment]
> While this probably won't help the author, I did a bisect. Starting from [b263005](https://github.com/dosbox-staging/dosbox-staging/commit/b26300533e052fc35ec8f72a224660742befe299) GBC is no longer compatible. Before that 32-bit works fine.

Well, that's another thing. We'll drop 32-bit builds anyway.

**Comment by johnnovak** [comment]
Closing as there is nothing we can do from our side to make this happen.

**Comment by Grounded0** [comment]
Changing close signal to can't fix.

**Comment by Grounded0** [comment]
If anyone wants to help out they should contact the author as DOSBox has had 5 years without a release and is now dead.

That should help out to get this moving.

**Comment by exoscoriae** [comment]
I've reached out to the author to seek a solution.

**Comment by johnnovak** [comment]
> GBC also currently only scans the lower memory, so 64-bit DOSBox has to be forced there

Yeah, we've killed the 32-bit builds a while ago and we're not bringing them back.

Nice find, btw 🎉

**Comment by johnnovak** [comment]
@emxd We tried the GOG version of Pool of Radiance with your `editbin` hack on the official 0.82.1 Staging executable, but no dice.

Can you please post some more details on how you managed to get this to work? Which Staging version you used, what config, which exact GOG release, etc. Thanks!

**Comment by johnnovak** [comment]
@emxd Thank you very much 👍🏻 

@Python-Exoproject Confirming the attached zip works.

![Image](https://github.com/user-attachments/assets/7fa3e0cf-50db-4287-99c6-086ea8e5dd95)

**Comment by Python-Exoproject** [comment]
I figured out why it didnt work for me when I tested it. Im trying to get it working in eXoDOS so I can roll the GBC launch option over to using Staging instead of ECE and one of the settings we use in eXoDOS for the game is to use Tandy sound as its three channel sound is superior to the pc speaker sound.

If you set the dosbox.conf in the files provided to use "pcspeaker=off" and "tandy=on", then use this Pool of Radiance config file ([POOL_CFG.zip](https://github.com/user-attachments/files/19720037/POOL_CFG.zip)) you will get the Tandy sound but most likely find that GBC wont detect anymore.

Im stunned at how sensitive this thing is, just using Tandy sound over PC Speaker makes it fail. Is there any tweaking to the method used here that might allow the Tandy sound to be used successfully? Otherwise for eXoDOS I will have to choose between ECE with better sound vs Staging with CRT autoshader and PC Speaker

**Comment by johnnovak** [comment]
> Im stunned at how sensitive this thing is, just using Tandy sound over PC Speaker makes it fail. Is there any tweaking to the method used here that might allow the Tandy sound to be used successfully? Otherwise for eXoDOS I will have to choose between ECE with better sound vs Staging with CRT autoshader and PC Speaker

It's regrettable, indeed. I don't imagine the GBC author would need to do anything overly difficult to adjust his thing a bit to accomodate Staging, but he doesn't seem to be interested in doing this.

Maybe @emxd has some further insights on the matter.

**Comment by johnnovak** [comment]
Re-opened and created a feature request ticket to make this work:
- https://github.com/dosbox-staging/dosbox-staging/issues/4277
